libbpf-tools/tcpconnect.c

```diff
@@ -0,0 +1,419 @@
+	int max_size = *size;
+	char *end;
+	long val;
+
+	for (*size = 0; *size < max_size; ) {
+		errno = 0;
+		val = strtol(argp, &end, 10);
+		if (errno) {
+			warn("strtol: %s: %s\n", arg, strerror(errno));
+			return -1;
+		} else if (end == arg || val < min || val > max) {
+			return -1;
+		}
+		ret[*size] = val;
+		*size += 1;
+		if (*end == 0)
+			break;
+		argp = end+1;
+	}
+	return 0;
+}
+
+static int get_uint(const char *arg, unsigned int *ret,
+		    unsigned int min, unsigned int max)
+{
+	char *end;
+	long val;
+
+	errno = 0;
+	val = strtoul(arg, &end, 10);
+	if (errno) {
+		warn("strtoul: %s: %s\n", arg, strerror(errno));
+		return -1;
+	} else if (end == arg || val < min || val > max) {
+		return -1;
+	}
+	if (ret)
+		*ret = val;
+	return 0;
+}
+
+static const struct argp_option opts[] = {
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ "timestamp", 't', NULL, 0, "Include timestamp on output" },
+	{ "count", 'c', NULL, 0, "Count connects per src ip and dst ip/port" },
+	{ "print-uid", 'U', NULL, 0, "Include UID on output" },
+	{ "pid", 'p', "PID", 0, "Process PID to trace" },
+	{ "uid", 'u', "UID", 0, "Process UID to trace" },
+	{ "port", 'P', "PORTS", 0,
+	  "Comma-separated list of destination ports to trace" },
+	{ "cgroupmap", 'C', "PATH", 0, "trace cgroups in this map" },
+	{ "mntnsmap", 'M', "PATH", 0, "trace mount namespaces in this map" },
+	{},
+};
+
+static struct env {
+	bool verbose;
+	bool count;
+	bool print_timestamp;
+	bool print_uid;
+	pid_t pid;
+	uid_t uid;
+	int nports;
+	int ports[MAX_PORTS];
+} env = {
+	.uid = (uid_t) -1,
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	int err;
+	int nports;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'c':
+		env.count = true;
+		break;
+	case 't':
+		env.print_timestamp = true;
+		break;
+	case 'U':
+		env.print_uid = true;
+		break;
+	case 'p':
+		err = get_int(arg, &env.pid, 1, INT_MAX);
+		if (err) {
+			warn("invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'u':
+		err = get_uint(arg, &env.uid, 0, (uid_t) -2);
+		if (err) {
+			warn("invalid UID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'P':
+		nports = MAX_PORTS;
+		err = get_ints(arg, &nports, env.ports, 1, 65535);
+		if (err) {
+			warn("invalid PORT_LIST: %s\n", arg);
+			argp_usage(state);
+		}
+		env.nports = nports;
+		break;
+	case 'C':
+		warn("not implemented: --cgroupmap");
+		break;
+	case 'M':
+		warn("not implemented: --mntnsmap");
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static volatile sig_atomic_t hang_on = 1;
+
+static void sig_int(int signo)
+{
+	hang_on = 0;
+}
+
+static void print_count_ipv4(int map_fd)
+{
+	struct ipv4_flow_key keys[MAX_ENTRIES];
```

Comment:
this can be static, no?

Reply:
same in ipv6 print

Reply:
Yes, replaced with static. But I wonder, what's the difference? We have plenty of stack in userspace

Reply:
don't know, frugality :) always tend to keep intermediate large buffer that are only used by one thread at a time as static. So I'd say stylistic preference.

Reply:
For per-thread I would also add `__thread` :)